Move handing of mavlink pre-flight calibration up into GCS_MAVLINK

peterbarker opened the pull request.

WickedShell commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -2089,6 +2089,81 @@ MAV_RESULT GCS_MAVLINK::handle_command_preflight_set_sensor_offsets(const mavlin
+    gcs().send_text(MAV_SEVERITY_INFO, "Updating barometer calibration");
+    AP::baro().update_calibration();
+    gcs().send_text(MAV_SEVERITY_INFO, "Barometer calibration complete");
+
```
Why not use the Airspeed singleton here and calibrate the airspeed sensor if it's not a nullptr? It would allow you to remove the virtual method/handling on baro from plane, as well as being ready to accept an airspeed sensor on any other vehicle if they choose to have them at a later point in time.

WickedShell commented:
I've only loosely skimmed this (not a full review), but I like that it catches the state of not being armed at the top level. I'm a bit taken aback you could get to doing a simple accel cal while armed (or flying) before on plane.

peterbarker commented:
On Mon, 19 Mar 2018, WickedShell wrote:

> Why not use the Airspeed singleton here and calibrate the airspeed sensor if

We have an airspeed singleton? :-)

WickedShell commented:
@peterbarker shoot it's presented in #7837 which is stalled for some reason, I thought that was merged already...

OXINARF commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -2089,6 +2089,81 @@ MAV_RESULT GCS_MAVLINK::handle_command_preflight_set_sensor_offsets(const mavlin
+MAV_RESULT GCS_MAVLINK::_handle_command_preflight_calibration_baro()
+{
+    gcs().send_text(MAV_SEVERITY_INFO, "Updating barometer calibration");
+    AP::baro().update_calibration();
```
Sub didn't do this, it did a full calibration. It's already overriding this method but still calling this.

peterbarker commented:
On Wed, 21 Mar 2018, Francisco Ferreira wrote:

> +    if (is_equal(packet.param5,4.0f)) {
> +        // simple accel calibration not supported for Copter
> What? Why? It was supported before...

Fixed.

> +MAV_RESULT GCS_MAVLINK_Tracker::_handle_command_preflight_calibration(const mavlink_command_long_t &packet)
> +{
> +    if (is_equal(packet.param4,1.0f)) {
> +        // Can't trim radio
> +        return MAV_RESULT_UNSUPPORTED;
> 
> Why is this needed? It was removed from other vehicles.

It got re-added accidentally; an earlier patch removed it.  I've removed 
this entire function from Tracker now, which is nice :-)

> > @@ -714,11 +714,6 @@ void GCS_MAVLINK_Copter::send_banner()
> Don't leave a blank line please.

Fixed.

> +    MAV_RESULT _handle_command_preflight_calibration_baro();
> Missing override.

Fixed.

> -    init_barometer(false);
> +    barometer.calibrate();
> 
> This is missing a false as an argument to the method, otherwise this is a functional change.

Added false.

> > @@ -645,11 +645,6 @@ void GCS_MAVLINK_Rover::handle_change_alt_request(AP_Mission::Mission_Command &c
>  MAV_RESULT GCS_MAVLINK_Rover::_handle_command_preflight_calibration(const mavlink_command_long_t &packet)
> No blank line please.

Fixed.


> +MAV_RESULT GCS_MAVLINK::_handle_command_preflight_calibration_baro()
> +{
> +    gcs().send_text(MAV_SEVERITY_INFO, "Updating barometer calibration");
> +    AP::baro().update_calibration();
> 
> Sub didn't do this, it did a full calibration. It's already overriding this method but still calling this.

Fixed.

This has really highlighted that Sub is weird when it comes to baro :-)

I've force-pushed this up.

peterbarker commented:
@WickedShell Can't use the airspeed singleton.  e.g. we don't enable it if plane says it isn't enabled.

*and* the calibration isn't a trivial call on the airspeed singleton.

No scope creeping this :-P

WickedShell commented:
@peterbarker but we just ask the library if it says it's enabled. I won't object to much, but I'm genuinely not seeing why it would be a large issue? Isn't it just:

```
AP_Airspeed *airspeed = AP_Airspeed.get_singleton();
if (airspeed != nullptr && airspeed->enabled()) {
    airspeed->calibrate(false);
    gcs().send_text(MAV_SEVERITY_INFO,"Airspeed calibration started");
}
```

If you really want you can even keep the baro update call. The read airspeed stuff is pointless. That's a legacy of the calibration being blocking and it not worth doing now. And even the baro cal can be skipped as all it was doing was updating the ground pressure/temp which you just did by doing the barometer calibration. I swear it's a small bit to enable the airspeed sensor!

OXINARF commented:
> This has really highlighted that Sub is weird when it comes to baro :-)

My guess is that it's because Sub can be armed while submerged.

OXINARF commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -714,6 +714,17 @@ void GCS_MAVLINK_Copter::handle_command_ack(const mavlink_message_t* msg)
 
+MAV_RESULT GCS_MAVLINK_Copter::_handle_command_preflight_calibration(const mavlink_command_long_t &packet)
+{
+
```
Don't leave a blank line please 🙏 😄

OXINARF commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -2219,6 +2219,81 @@ MAV_RESULT GCS_MAVLINK::handle_command_preflight_set_sensor_offsets(const mavlin
+    }
+
+    if (is_equal(packet.param3,1.0f)) {
+        // fast barometer calibration
```
Move this comment to inside the method, since the method is overridden by Sub that does a full calibration.

jaxxzer commented:
> My guess is that it's because Sub can be armed while submerged.

Exactly. Subs can be armed/disarmed at any arbitrary point, even at depth. We need to be sure that the baro is only calibrated at the surface.

peterbarker commented:
On Wed, 28 Mar 2018, Francisco Ferreira wrote:

>  *  Commit Rover: factor out preflight calibration function says factor out,
>     but it isn't a direct copy, stuff is missing

Changed the commit message and added a note that some stuff had moved up


>  *  Commit GCS_MAVLink: tidy handling of barometer calibrations should
>     probably be before the vehicle ones but I don't care much

Reordering isn't fun with this sort of change :-P


> mavlink_command_long_t &packet)
> +{
> +
> 
> Don't leave a blank line please 🙏 😄

Fixed.

> Move this comment to inside the method, since the method is overridden by
> Sub that does a full calibration.

Good point!

Force-pushed up.

OXINARF commented:
@peterbarker 

> Changed the commit message and added a note that some stuff had moved up

Not great that only Rover does that, but whatever.

> Reordering isn't fun with this sort of change :-P

Should have been easy as each commit is split by library (that's the biggest advantage of our system). But whatever 😄

peterbarker commented:
On Mon, 2 Apr 2018, Francisco Ferreira wrote:

>       Reordering isn't fun with this sort of change :-P
> 
> Should have been easy as each commit is split by library (that's the biggest
> advantage of our system). But whatever 😄

It's more reordering the bits in the same library.

OXINARF commented:
> It's more reordering the bits in the same library.

Ok, but I didn't ask for that 😉 It was just moving one GCS_MAVLink commit to be before the **corresponding** vehicle changes.

Merged together with #8024.
